Pinta/MainWindow.cs

```diff
@@ -163,47 +165,127 @@ private void DockNotebook_TabClosed (object sender, TabClosedEventArgs e)
+            var view = (DocumentViewContent)content.ViewContent;
+
+            PintaCore.Workspace.SetActiveDocument (view.Document);
+
+            ((CanvasWindow)view.Control).Canvas.GdkWindow.Cursor = PintaCore.Tools.CurrentTool.CurrentCursor;
+        }
+
         private void DockNotebook_ActiveTabChanged (object sender, EventArgs e)
         {
-            var item = canvas_pad.Notebook.ActiveItem;
+            var tab = DockNotebookManager.ActiveTab;
 
-            if (item == null)
+            if (tab == null || tab.Content == null)
                 return;
 
-			var view = (DocumentViewContent)item;
+            var content = (SdiWorkspaceWindow)tab.Content;
+            var view = (DocumentViewContent)content.ViewContent;
 
             PintaCore.Workspace.SetActiveDocument (view.Document);
 
-            ((CanvasWindow)view.Widget).Canvas.Window.Cursor = PintaCore.Tools.CurrentTool.CurrentCursor;
+            ((CanvasWindow)view.Control).Canvas.GdkWindow.Cursor = PintaCore.Tools.CurrentTool.CurrentCursor;
+        }
+
+        private void Workspace_DocumentClosed (object sender, DocumentEventArgs e)
+        {
+            var tab = FindTabWithCanvas ((PintaCanvas)e.Document.Workspace.Canvas);
+
+            if (tab != null)
+                dock_container.CloseTab (tab);
+        }
+
+        private void MainWindow_DragDataReceived (object o, DragDataReceivedArgs args)
+        {
+	        // TODO: Generate random name for the picture being downloaded
+
+	        // Only handle URIs
+	        if (args.Info != 100)
+		        return;
+
+	        string fullData = System.Text.Encoding.UTF8.GetString (args.SelectionData.Data);
+
+	        foreach (string individualFile in fullData.Split ('\n')) {
+		        string file = individualFile.Trim ();
+
+		        if (file.StartsWith ("http") || file.StartsWith ("ftp")) {
+			        var client = new System.Net.WebClient ();
+			        string tempFilePath = System.IO.Path.GetTempPath () + System.IO.Path.GetFileName (file);
+
+			        var progressDialog = PintaCore.Chrome.ProgressDialog;
+
+			        try {
+				        PintaCore.Chrome.MainWindowBusy = true;
+
+				        progressDialog.Title =  ("Downloading Image");
+				        progressDialog.Text = "";
+				        progressDialog.Show ();
+
+				        client.DownloadProgressChanged += (sender, e) => {
+					        progressDialog.Progress = e.ProgressPercentage;
+				        };
+
+				        client.DownloadFile (file, tempFilePath);
+
+				        if (PintaCore.Workspace.OpenFile (tempFilePath)) {
+					        // Mark as not having a file, so that the user doesn't unintentionally
+					        // save using the temp file.
+					        PintaCore.Workspace.ActiveDocument.HasFile = false;
+				        }
+			        } catch (Exception e) {
+				        progressDialog.Hide ();
+				        PintaCore.Chrome.ShowErrorDialog (PintaCore.Chrome.MainWindow,
+					         ("Download failed"),
```

Comment:
Could you replace this function with the current version from the gtk3-v2 branch? For example, the call to Translations.GetString() is missing so the string here won't be translated.